modules/standard/Version.chpl

```diff
@@ -0,0 +1,349 @@
+    return "version " + x.major:string + "." + x.minor:string + "." +
+           x.update:string;
+  }
+
+
+  /*
+    This record represents a software version in a Git repository.  It
+    uses ``param`` values to represent its components in order to
+    support compile-time comparison of version numbers which in turn
+    permits code to specialize to specific versions of Chapel.  When
+    printed or converted to a string, it is represented as ``version
+    major.minor.update (commit)``.
+
+    Note that comparisons between two sourceVersion values with identical
+    semantic versions are not supported due to the challenges involved
+    in ordering commit values.  In addition, when a semanticVersion is
+    compared with a sourceVersion with the identical semantic version,
+    the sourceVersion value is considered "less than" the semanticVersion
+    using the model that it represents a pre-release of the official
+    release represented by the semantic version.
+  */
+  record sourceVersion {
+    forwarding version only major, minor, update;
+
+    // TODO: Why does the following type declaration break compilation?
+    //    const version /*: semanticVersion(?,?,?) */;
+
+    /* The semantic version portion of the version number. */
+    const version;
+
+    /* The commit ID of the version (e.g., a git SHA) */
+    param commit: string;
+
+    pragma "no doc"
+    proc init(param major: int, param minor: int, param update: int,
+              param commit: string) {
+      this.version = new semanticVersion(major, minor, update);
+      this.commit = commit;
+      if commit == "" then
+        compilerError("sourceVersion values must have a commit value");
+    }
+
+    pragma "no doc"
+    proc writeThis(s) throws {
+      s.write(this:string);
+    }
+  }
+
+  // cast from sourceVersion to string
+  pragma "no doc"
+  proc _cast(type t: string, x: sourceVersion(?)) {
+    return x.version:string + " (" + x.commit + ")";
+  }
+
+
+  // helper functions for comparison operators
+
+  private proc spaceship(param x: int, param y: int) param : int {
+    if x < y then
+      return -1;
+    else if x == y then
+      return 0;
+    else
+      return 1;
+  }
+
+  private proc spaceship(v1: semanticVersion(?),
+                         v2: semanticVersion(?)) param : int {
+    param majComp = spaceship(v1.major, v2.major);
+    if majComp != 0 then
+      return majComp;
+    param minComp = spaceship(v1.minor, v2.minor);
+    if minComp != 0 then
+      return minComp;
+    return spaceship(v1.update, v2.update);
+  }
+
+
+  // Comparisons between semanticVersions
+
+  proc ==(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
+    return spaceship(v1,v2) == 0;
+  }
+
+  proc !=(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
+    return spaceship(v1,v2) != 0;
+  }
+
+  proc <(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
+    return spaceship(v1,v2) == -1;
```

Comment:
I personally find these easier to think about/parse when written in terms of comparisons to 0:

 - `proc ==` -> `== 0`
 - `proc < `-> `< 0`
 - `proc <=` -> `<= 0`

I'm not sure it's worth changing since this is tested, I just have a slow time looking at these and convincing myself they're right.

Reply:
Good suggestion.  I'm not familiar enough with the spaceship operator to have thought of this pattern.